src/journal/journal.c

```diff
@@ -0,0 +1,678 @@
+}
+
+/* Exported function */
+TelemJournal *open_journal(const char *journal_file)
+{
+        FILE *fptr = NULL;
+        char *boot_id = NULL;
+        struct TelemJournal *telem_journal;
+
+        // Use default location if journal_file parameter is NULL
+        fptr = (journal_file == NULL) ? fopen(JOURNAL_PATH, "a+") :
+               fopen(journal_file, "a+");
+
+        if (fptr == NULL) {
+#ifdef DEBUG
+                perror("Error: ");
+#endif
+                return NULL;
+        }
+
+        if (read_boot_id(&boot_id) != 0) {
+#ifdef DEBUG
+                perror("Error: ");
+#endif
+                return NULL;
+        }
+
+        telem_journal = malloc(sizeof(struct TelemJournal));
+        if (!telem_journal) {
+                fprintf(stderr, "Unable to allocate more memory\n");
+                return NULL;
+        }
+
+        telem_journal->fptr = fptr;
+        telem_journal->journal_file = \
+                (journal_file == NULL) ? strdup(JOURNAL_PATH) : strdup(journal_file);
+        telem_journal->boot_id = boot_id;
+        telem_journal->record_count = get_record_count(telem_journal);
+        telem_journal->record_count_limit = RECORD_LIMIT;
+
+#ifdef DEBUG
+        printf("Records in db: %d\n", telem_journal->record_count);
+#endif
+
+        return telem_journal;
+}
+
+/* Exported function */
+void close_journal(TelemJournal *telem_journal)
+{
+	if (telem_journal) {
+                free(telem_journal->boot_id);
+                free(telem_journal->journal_file);
+                fclose(telem_journal->fptr);
+                free(telem_journal);
+        }
+}
+
+/* Exported function */
+int print_journal(TelemJournal *telem_journal, char *classification,
+                   char *record_id, char *event_id, char *boot_id)
+{
+        int n = 0;
+        int rc = 0;
+        int count = 0;
+        char str_time[80] = {0};
+        char *line = NULL;
+        size_t len = 0;
+        FILE *journal_fileptr = NULL;
+        struct tm ts;
+        struct JournalEntry *entry = NULL;
+
+        if (telem_journal == NULL) {
+                return -1;
+        }
+
+        journal_fileptr = fopen(telem_journal->journal_file, "r");
+        if (!journal_fileptr) {
+                return -1;
+        }
+
+        // Advance file pointer to line n
+        n = telem_journal->record_count - telem_journal->record_count_limit;
+        if (n > 0) {
+                rc = skip_n_lines(n, journal_fileptr);
+                if (rc == -1) {
+                        return rc;
+                } else if (rc != 0) {
+                        fprintf(stderr, "Error: ", strerror(rc));
+                }
+        }
+
+        while (getline(&line, &len, journal_fileptr) != -1) {
+                deserialize_journal_entry(line, &entry);
+                if (entry) {
+                        /* filter entry out if one is provided */
+                        if (record_id != NULL && strncmp(entry->record_id, record_id, ID_LEN) != 0) {
```

Comment:
Not clear if this is trying to see if the records/classification/boot_id/event_id are prefixes or equal.

If the idea is to test equal, then just use strcmp rather than strncmp.

If the idea is to use prefix, then you probably want to use the length of the filtering value (i.e. boot_id, event_id ...)